src/compiler/checker.ts

```diff
@@ -16535,6 +16517,89 @@ namespace ts {
+                return createUnionOfSignaturesForOverloadFailure(candidates);
+            }
+
+            // Pick the longest signature. This way we can get a contextual type for cases like:
+            //     declare function f(a: { xa: number; xb: number; }, b: number);
+            //     f({ |
+            // Also, use explicitly-supplied type arguments if they are provided, so we can get a contextual signature in cases like:
+            //     declare function f<T>(k: keyof T);
+            //     f<Foo>("
+            const bestIndex = getLongestCandidateIndex(candidates, apparentArgumentCount === undefined ? args.length : apparentArgumentCount);
+            const candidate = candidates[bestIndex];
+
+            const { typeParameters } = candidate;
+            if (typeParameters && callLikeExpressionMayHaveTypeArguments(node) && node.typeArguments) {
+                const typeArguments = node.typeArguments.map(getTypeOfNode);
+                while (typeArguments.length > typeParameters.length) {
+                    typeArguments.pop();
+                }
+                while (typeArguments.length < typeParameters.length) {
+                    typeArguments.push(getDefaultTypeArgumentType(isInJavaScriptFile(node)));
+                }
+
+                const instantiated = createSignatureInstantiation(candidate, typeArguments);
+                candidates[bestIndex] = instantiated;
+                return instantiated;
+            }
+
+            return candidate;
+        }
+
+        function createUnionOfSignaturesForOverloadFailure(candidates: ReadonlyArray<Signature>): Signature {
+            const thisParameters = mapDefined(candidates, c => c.thisParameter);
+            let thisParameter: Symbol | undefined;
+            if (thisParameters.length) {
+                thisParameter = createCombinedSymbolWithType(thisParameters, getUnionType(thisParameters.map(getTypeOfParameter), /*subtypeReduction*/ true));
+            }
+
+            const { min: minArgumentCount, max: maxNonRestParam }  = minAndMax(candidates, getNumNonRestParameters);
+            const hasRestParameter = candidates.some(c => c.hasRestParameter);
+            const hasLiteralTypes = candidates.some(c => c.hasLiteralTypes);
+            const parameters: ts.Symbol[] = [];
+            for (let i = 0; i < maxNonRestParam; i++) {
+                const symbols = mapDefined(candidates, ({ parameters, hasRestParameter }) => hasRestParameter ?
+                    i < parameters.length - 1 ? parameters[i] : last(parameters) :
+                    i < parameters.length ? parameters[i] : undefined);
+                Debug.assert(symbols.length !== 0);
+                const types = mapDefined(candidates, candidate => tryGetTypeAtPosition(candidate, i));
+                parameters.push(createCombinedSymbolWithType(symbols, getUnionType(types, /*subtypeReduction*/ true)));

```

Comment:
instead of creating a small-but-already-leaky abstraction, what about creating a single-use createCombinedSymbol that accounts for other common code, like the calls to [try]GetTypeOf[Parameter/Rest], mapDefined and getUnionType.

That function would not misrepresent itself in its name, and would make this function considerably more readable.